src/source_info.rs

```diff
@@ -304,22 +396,83 @@ impl<'i> SourceInfoBuilder<'i> {
         result
     }
 
+    fn get_definition_from_frame(
+        &self,
+        id: &str,
+        frame_id: FrameId,
+        location: &Location,
+    ) -> Option<&Definition> {
+        let mut current_frame_id = Some(frame_id);
+        while let Some(fid) = current_frame_id {
+            // Search definitions in the current scope
+            for definition in self
+                .definitions
+                .iter()
+                .rev()
+                .chain(self.imported_definitions.iter())
+            {
+                let ordering = cmp_range_to_range(&definition.location.range, location.range);
+                if definition.frame_id == fid
+                    && definition.id.as_str() == id
+                    && (ordering == Ordering::Equal || ordering == Ordering::Less)
+                {
+                    return Some(definition);
+                }
+            }
+            // Move to parent scope if available
+            current_frame_id = self.frames.get(fid.0).and_then(|frame| frame.parent);
+        }
+        None
+    }
+
+    fn resolve_references(&mut self) {
+        for unresolved in &self.unresolved_references {
+            let id = unresolved.id.as_str();
+            let frame_id = unresolved.frame_id;
+            let Some(def) = self.get_definition_from_frame(id, frame_id, &unresolved.location)
+            else {
+                // Definition not found
+                continue;
+            };
+
+            self.references.push(Reference {
+                location: unresolved.location.clone(),
+                definition: def.location.clone(),
+                id: unresolved.id.clone(),
+                frame_id: unresolved.frame_id,
+            });
+        }
+    }
+
     fn build(mut self, error: Option<Error>) -> SourceInfo {
-        // Sort the definitions, they get added in pop_frame so they aren't in order
+        // Resolve unresolved references
+        self.resolve_references();
+
+        self.references
+            .sort_by_key(|reference| reference.location.range.start);
+
         self.definitions
             .sort_by_key(|definition| definition.location.range.start);
 
-        // References should already be sorted
+         // References should already be sorted
         debug_assert!(is_sorted::IsSorted::is_sorted_by_key(
             &mut self.references.iter(),
             |reference| reference.location.range.start
         ));
 
+        // Definitions should already be sorted
+        debug_assert!(is_sorted::IsSorted::is_sorted_by_key(
+            &mut self.definitions.iter(),
+            |definition| definition.location.range.start
+        ));
+
```

Comment:
The assertion for `self.references` was there because `self.definitions` was being sorted, so it made sense to demonstrate that sorting the references was unnecessary. I think that now that the references also get sorted, the assertions here are unnecessary, and are only really validating that `sort_by_key` works as expected.